domain/upgrade/state/state.go

```diff
@@ -0,0 +1,311 @@
+			return errors.Annotatef(err, "preparing %q", q)
+		}
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{
+			"uuid":             upgradeUUID,
+			"previous_version": previousVersion.String(),
+			"target_version":   targetVersion.String(),
+		}).Run())
+	})
+
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return upgradeUUID, nil
+}
+
+// SetControllerReady marks the supplied controllerID as being ready
+// to start a provided upgrade. All provisioned controllers need to
+// be ready before an upgrade can start
+// A controller node is ready for an upgrade if a row corresponding
+// to the controller is present in upgrade_info_controller_node
+func (st *State) SetControllerReady(ctx context.Context, upgradeUUID, controllerID string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		nodeInfos, err := st.getUpgradeNodes(ctx, tx, upgradeUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, nodeinfo := range nodeInfos {
+			if nodeinfo.ControllerNodeID == controllerID {
+				return nil
+			}
+		}
+		q := `
+INSERT INTO upgrade_info_controller_node (uuid, controller_node_id, upgrade_info_uuid) VALUES
+($M.uuid, $M.controller_id, $M.info_uuid)`
+		s, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Annotatef(err, "preparing %q", q)
+		}
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{
+			"uuid":          utils.MustNewUUID().String(),
+			"controller_id": controllerID,
+			"info_uuid":     upgradeUUID,
+		}).Run())
+	}))
+}
+
+// AllProvisionedControllersReady returns true if and only if all controllers
+// that have been started by the provisioner are ready to start the provided upgrade
+func (st *State) AllProvisionedControllersReady(ctx context.Context, upgradeUUID string) (bool, error) {
+	db, err := st.DB()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	var (
+		provisionedControllers []string
+		nodeInfos              []infoControllerNode
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var err error
+		provisionedControllers, err = st.getProvisionedControllers(ctx, tx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		nodeInfos, err = st.getUpgradeNodes(ctx, tx, upgradeUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	ready := set.NewStrings(transform.Slice(nodeInfos, func(info infoControllerNode) string { return info.ControllerNodeID })...)
+	missing := set.NewStrings(provisionedControllers...).Difference(ready)
+	return missing.IsEmpty(), nil
+}
+
+func (st *State) getProvisionedControllers(ctx context.Context, tx *sqlair.TX) ([]string, error) {
+	q := `SELECT controller_id FROM controller_node WHERE dqlite_node_id IS NOT NULL`
+	s, err := sqlair.Prepare(q)
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
```

Comment:
We're creating these every iteration. I'm not sure how wise this is. Maybe think about moving this somewhere else by passing the sqlair statement to the function.